lib/puppet/defaults.rb

```diff
@@ -727,13 +727,35 @@ def self.initialize_default_settings!(settings)
         unless [0, 'unlimited', Float::INFINITY].include?(val)
           Puppet.deprecation_warning(<<-WARNING)
 Fine grained control of environment timeouts is deprecated,
-please use `0` or `unlimited` to control default caching behavior
-and the environment-cache endpoint in Puppet Server's administrative
-API to expire the cache as needed
+please use `environment_ttl` instead.
           WARNING
         end
       end
     },
+    :environment_ttl => {
+      :default    => "0",
+      :type       => :ttl,
+      :desc       => "How long the puppet server should cache data it loads from an
+      environment based on when it was last accessed.
+
+      A value of `0` will disable caching. This setting can also be set to
+      `unlimited`, which will cache environments until the server is restarted
+      or told to refresh the cache. All other values will result in the server
+      evicting environments that have not been accessed within that ttl.
+
+      You should change this setting once your Puppet deployment is doing
+      non-trivial work. We chose the default value of `0` because it lets new
+      users update their code without any extra steps, but it lowers the
+      performance of your Puppet server.
+
+      We recommend setting this to `unlimited` and explicitly refreshing your
+      Puppet master as part of your code deployment process.
```

Comment:
This looks like a copy/paste from `environment_timeout` but I believe is incorrect.  

"We recommend setting this to a number that will keep your most actively used environments cached but allow testing environments to fall out of the cache and reduce memory usage."  Or some such?

I was considering a reasonable suggested default might be 3 minutes.

Reply:
Yeah, it was very much copy/pasta'd. I have no idea what a good default value might be... Happy to change it to 3 minutes if that seems right. Thoughts @Sharpie?

Reply:
@npwalker I updated the text to what you recommended and mentioned that environments can still be manually evicted using the `environment-cache` API.

About the default value, this change is going into 6.x and 7. I wanted to avoid having different defaults across those branches. If we make the default `3m`, then it's surprising to users getting started in 7 (due to there not being an `environment_timeout`). So I kept the default `0` and made a note recommending `3m` instead of `unlimited`. Let me know what you think?

Reply:
yea that's better.  I agree we want the default to stay at 0.  I was recommending `3m` as a good choice when you are configuring it.  So this looks right to me.  

How often is the cache actually checked against the TTL?  If I set TTL to `5s` what's the actual behavior?

Reply:
The behavior is whenever an environment is requested via `Puppet::Environments#get(:name)`, we check each cached environment to see if it was last accessed within the past `5s`. If not, it's evicted. So setting a low TTL doesn't cause us to check more frequently. Instead it shortens the window of time that we'll consider an environment to be "still used". If the ttl is too short, eg shorter than the amount of time it takes to compile the catalog, then it's possible we'll evict environments unnecessarily.  Puppet's debug output (and that of puppetserver) will print if environments are being evicted or reused, so folks can tune the TTL value. I'm not sure if we have metrics around environment cache hits/misses, if not we might want to add that at some point.